chapter.md

```diff
@@ -0,0 +1,966 @@
+  payload: { path, value }
+});
+
+const putInArray = (path, value, index) => ({
+  type: 'UPDATE_INDEXED_FORM_VALUE',
+  payload: { path, value, index }
+});
+
+const removeFromArray = (index, path) => ({
+  type: 'REMOVE_INDEXED_FORM_VALUE',
+  payload: { index, path }
+});
+```
+
+There's nothing exciting going on here:
+each function creates an action with a `type` (again, required by Redux)
+and a `payload` (our own term).
+To cut down on typing,
+we make use of the fact that `{a, b}` is a shorthand for
+`{a: a, b: b}`:
+
+```js
+let a = 'A';
+let b = 'B';
+let both = {a, b};
+console.log(both);
+```
+```
+{a: 'A', b: 'B'}
+```
+
+The payload of the action returned by `addIntoArray`
+therefore has two keys `path` and `value`,
+which are in turn bound to whatever was passed in
+for the parameters with the corresponding names.
+
+Once we have these actions,
+the next step is to update our reducer by adding the following case:
+
+```ts
+case 'ADD_INDEXED_FORM_VALUE':
+  const lensForProp = lensPath(action.payload.path);
+  const propValue = <any[]> view(lensForProp, state);
+  return assocPath(
+    action.payload.path,
+    concat(propValue, [action.payload.value]),
+    state
+  );
+```
+
+Again,
+we're using Ramda functions to create a new array of skills
+given an old array and the thing we want to add:
+`lensForProp` is the location of the skills array in our state,
+`propValue` is its current value,
+and `concat` wrapped up in `assocPath` gives us the old skills
+with the new one appended.
+The additions to handle updating and removing skills have a similar shape.
+
+Extending our user interface is equally easy.
+We start with a helper function that creates and dispatches the required action:
+
+```ts
+addSkill() {
+  this.ngRedux.dispatch(addIntoArray({
+    value: undefined,
+    path: [ 'character', 'skills' ]
+  }));
+}
+```
+
+We have put this in the form component,
+but neither Angular nor good design principles strictly require this:
+we could have put it (and similar functions) in a file full of utilities.
+
+Now that we have the functions we need,
+we can write the HTML needed to put everything in front of the user:
+
+```html
+<label>Skills:</label>
+
+<div *ngFor="let skillSlot of characterForm.skills; let i = index;">
+  <select
+    [value]="skillSlot"
+    (change)="onSelectSkill($event, i)">
+    <option *ngFor="let skill of skills" [value]="skill">
+      {{ skill }}
+    </option>
+  </select>
+  <button type="button" (click)="removeSkill(i)">Remove</button>
+</div>
+
+<button type="button" (click)="addSkill()">Add skill</button>
+```
+
+The last line of this HTML is the one that lets users add skills;
+the rest is to handle skill display and removal.
+
+## The Payoff: Validation
+
+One way to see how these changes pays off is to look at validation,
+and in particular at the way in which we can use *selectors* to compute data by composing functions,
+and then use *memoization* to make state changes more efficient.
```

Comment:
My concern here is mainly that selectors never change the state and we want to make sure we don't imply otherwise by mentioning state changes without being clear where selectors exist in the redux flow.